src/fetch/object_id.rs

```diff
@@ -60,6 +67,69 @@ where
     Kind: Object,
     for<'de2> <Kind as Object>::Kind: serde::Deserialize<'de2>;
 
+/// get a cache map for a specific object kind
+/// creates a new cache map if not already exists
+async fn get_cache_for_type<Kind>() -> Cache<Url, Result<Kind, Kind::Error>>
+where
+    Kind: Object + Send + Sync + Clone + 'static,
+    <Kind as Object>::Error: Clone + Send + Sync,
+{
+    static CACHES: OnceLock<RwLock<type_map::concurrent::TypeMap>> = OnceLock::new();
+    let caches = CACHES.get_or_init(|| RwLock::new(TypeMap::new()));
+    // in most cases (except the first call) we only need a read-only lock, do hot path separately for perf
+    let existing_cache = {
+        caches
+            .read()
+            .await
+            .get::<Cache<Url, Result<Kind, Kind::Error>>>()
+            .map(|e| e.clone())
+    };
+    match existing_cache {
+        Some(cache) => cache,
+        None => {
+            let mut caches = caches.write().await;
+            caches
+                .entry()
+                .or_insert_with(|| {
+                    Cache::builder()
+                        .max_capacity(Kind::cache_max_capacity())
+                        .time_to_live(Kind::cache_time_to_live())
+                        .build()
+                })
+                .clone()
+        }
+    }
+}
+impl<Kind> ObjectId<Kind>
+where
+    Kind: Object + Send + Sync + Clone + 'static,
+    for<'de2> <Kind as Object>::Kind: serde::Deserialize<'de2>,
+    <Kind as Object>::Error: Clone + Send + Sync,
+{
+    /// Fetches an activitypub object, either from local database (if possible), or over http, retrieving from cache if possible
+    #[tracing::instrument(skip(data))]
+    pub async fn dereference(
+        &self,
+        data: &Data<<Kind as Object>::DataType>,
+    ) -> Result<Kind, <Kind as Object>::Error>
+    where
+        <Kind as Object>::Error: From<Error> + From<anyhow::Error>,
+    {
+        if Kind::cache_max_capacity() == 0 || Kind::cache_time_to_live().is_zero() {
+            self.dereference_uncached(data).await
+        } else {
+            // The Cache.get_with method ensures that the dereference_inner method is only called once even if the dereference method is called twice simultaneously.
+            // From the moka docs: "This method guarantees that concurrent calls on the same not-existing key are coalesced into one evaluation of the init future. Only one of the calls evaluates its future, and other calls wait for that future to resolve."
+            // Considerations: should an error result be stored in the cache as well? Right now: yes. Otherwise try_get_with should be used.
```

Comment:
I dont think it makes sense to cache error values. I would expect errors to be temporary and retried the next time.

Reply:
> I dont think it makes sense to cache error values. I would expect errors to be temporary and retried the next time.

The main reason I cached them is that failed lookups are very common and "intentional" in lemmy. For example, when inserting a comment it first looks up if the corresponding parent exist, and if it doesn't it tries looking that up which might permanently fail. That lookup failing is a fairly hot path and sending out spamming http requests is not great.

Tbh, I can't think of good example right now where these errors are something that would happen "expectedly", maybe e.g. if some instance is blacklisted so their content is not inserted, but I think it does exist.

In general, I think lemmy needs to start distinguishing between ephemeral errors and permanent ones. Right now the code eats up lots of errors that should be passed up or ignored (e.g. parsing error vs 404 error vs DNS lookup error).